pkg/controller/atlasdeployment/atlasdeployment_controller.go

```diff
@@ -188,18 +179,38 @@ func (r *AtlasDeploymentReconciler) Reconcile(context context.Context, req ctrl.
 	}
 
 	handleDeployment := r.selectDeploymentHandler(deployment)
-	if result, _ := handleDeployment(workflowCtx, project, deployment, req); !result.IsOk() {
+	if result, _ := handleDeployment(context, workflowCtx, project, deployment, req); !result.IsOk() {
 		workflowCtx.SetConditionFromResult(status.DeploymentReadyType, result)
-		return result.ReconcileResult(), nil
+		return r.registerConfigAndReturn(workflowCtx, context, log, deployment, result), nil
 	}
 
 	if !deployment.IsServerless() {
 		if result := r.handleAdvancedOptions(workflowCtx, project, deployment); !result.IsOk() {
 			workflowCtx.SetConditionFromResult(status.DeploymentReadyType, result)
-			return result.ReconcileResult(), nil
+			return r.registerConfigAndReturn(workflowCtx, context, log, deployment, result), nil
+		}
+	}
+
+	return r.registerConfigAndReturn(workflowCtx, context, log, deployment, workflow.OK()), nil
+}
+
+func (r *AtlasDeploymentReconciler) registerConfigAndReturn(
+	workflowCtx *workflow.Context,
+	context context.Context,
+	log *zap.SugaredLogger,
+	deployment *mdbv1.AtlasDeployment,
+	result workflow.Result) ctrl.Result {
+	if result.IsOk() || result.IsInProgress() {
```

Comment:
cc @igor-karpukhin 
I think this was the issue before I was ONLY saving the last config if all went fine till the end of the reconcile method and was going to return OK. But there are times when the `handleDeployment` or `handleAdvancedOptions` might need to exit early but still it is not due to an abrupt failure or interruption of the reconcile loop, but just because the action is still "in progress" and this is has to be revisited later.

My reasoning here is that if the exit at those points is either `OK` or `In Progress` it means that the operator has already taken ownership of the Atlas counterpart and therefore we need to save the current config as applied, even if it is still being applied (in progress).

I hope to see a lot more green in the CI tests after this fix.